dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/security/AuthorizationData.java

```
@@ -0,0 +1,148 @@
+
+    /**
+     * Used to create the Authorization Data in the correct way
+     *
+     * If the security annotation is a:
+     * - class annotation -> get roles list from class annotation (the method's declaring type)
+     * - method annotation -> get roles list from method annotation (the method's declaring type)
+     *
+     * If there is no security annotation then return an empty role list
+     *
+     * @param m the annotated method
+     * @return
+     */
+    public static AuthorizationData createFrom(AnnotatedMethod<?> m) {
+        SecurityAnnotation securityAnnotation = collectSecurityAnnotation(m);
+        AnnotationType annotationType = ANNOTATION_TYPE_MAP.get(securityAnnotation);
+
+        List<String> computedRoleList = switch (annotationType) {
+            case CLASS -> {
+                yield getRoleList(m.getDeclaringType());
+            }
+            //get roles list from method annotation
+            case METHOD -> {
+                yield getRoleList(m);
+            }
+            //return empty role list
+            case NONE -> {
+                yield new ArrayList<>();
+            }
+            //return empty role list
+            default -> {
+                yield new ArrayList<>();
+            }
+        };
+
+        return new AuthorizationData(securityAnnotation, computedRoleList);
+    }
+
+    /**
+     *
+     * Return the rolelist for the annotation
+     *
+     * @param a the annotation - This can be a method or class annotation
+     * @return the roleList for the given annotation
+     */
+    private static List<String> getRoleList(Annotated a) {
+        List<String> roleList = new ArrayList<>();
+
+        if (a != null && a.isAnnotationPresent(RolesAllowed.class)) {
+            RolesAllowed annotation = a.getAnnotation(RolesAllowed.class);
+            Collections.addAll(roleList, annotation.value());
+        }
+        return roleList;
+    }
```

The `if` block here was added in case `a` is somehow `null` or doesn't have the expected annotation. I don't think that will ever happen with the current code but if it did, that would be an error and we should probably throw an exception.

However, this code to extract the role list could be moved into `collectSecurityAnnotation` which is doing the search for the annotations. Once you've found the annotation you can extract the role list, but since you already have the annotation, you don't need to check it exists first.